mm2src/common/expirable_map.rs

```diff
@@ -19,48 +19,80 @@ impl<V> ExpirableEntry<V> {
 /// WARNING: This is designed for performance-oriented use-cases utilizing `FxHashMap`
 /// under the hood and is not suitable for cryptographic purposes.
 #[derive(Clone, Debug)]
-pub struct ExpirableMap<K: Eq + Hash, V>(FxHashMap<K, ExpirableEntry<V>>);
+pub struct ExpirableMap<K: Eq + Hash + Copy, V> {
+    map: FxHashMap<K, ExpirableEntry<V>>,
+    /// A sorted inverse map from expiration times to keys to speed up expired entries clearing.
+    expiries: BTreeMap<Instant, K>,
+}
 
-impl<K: Eq + Hash, V> ExpirableMap<K, V> {
+impl<K: Eq + Hash + Copy, V> ExpirableMap<K, V> {
     /// Creates a new empty `ExpirableMap`
     #[inline]
-    pub fn new() -> Self { Self(FxHashMap::default()) }
+    pub fn new() -> Self {
+        Self {
+            map: FxHashMap::default(),
+            expiries: BTreeMap::new(),
+        }
+    }
 
     /// Returns the associated value if present.
+    ///
+    /// Note that if the entry is expired and wasn't cleared yet, it will not be returned.
+    /// Use `remove()` instead to get expired entries if they still exist.
+    #[inline]
+    pub fn get(&self, k: &K) -> Option<&V> {
+        self.map
+            .get(k)
+            .filter(|v| v.expires_at > Instant::now())
+            .map(|v| &v.value)
+    }
+
+    /// Removes a key-value pair from the map and returns the associated value if present.
+    ///
+    /// Note that if the entry is expired and wasn't cleared yet, it will be returned.
     #[inline]
-    pub fn get(&mut self, k: &K) -> Option<&V> { self.0.get(k).map(|v| &v.value) }
+    pub fn remove(&mut self, k: &K) -> Option<V> {
+        let entry = self.map.remove(k)?;
+        self.expiries.remove(&entry.expires_at);
+        Some(entry.value)
+    }
 
     /// Inserts a key-value pair with an expiration duration.
     ///
     /// If a value already exists for the given key, it will be updated and then
     /// the old one will be returned.
     pub fn insert(&mut self, k: K, v: V, exp: Duration) -> Option<V> {
-        let entry = ExpirableEntry {
-            expires_at: Instant::now() + exp,
-            value: v,
-        };
-
-        self.0.insert(k, entry).map(|v| v.value)
+        self.clear_expired_entries();
+        let expires_at = Instant::now() + exp;
+        let entry = ExpirableEntry { expires_at, value: v };
+        self.expiries.insert(expires_at, k);
+        self.map.insert(k, entry).map(|v| v.value)
     }
 
     /// Removes expired entries from the map.
-    pub fn clear_expired_entries(&mut self) { self.0.retain(|_k, v| Instant::now() < v.expires_at); }
-
-    /// Removes a key-value pair from the map and returns the associated value if present.
-    #[inline]
-    pub fn remove(&mut self, k: &K) -> Option<V> { self.0.remove(k).map(|v| v.value) }
+    fn clear_expired_entries(&mut self) {
+        let now = Instant::now();
+
+        while let Some((exp, key)) = self.expiries.pop_first() {
+            if exp > now {
+                self.expiries.insert(exp, key);
```

Comment:
I think `first_key_value` should improve the performance here. For non-expired entries (which means for a lot of entries) this has overhead with removing and inserting back the same value.

Reply:
> With pop_first removing and inserting back will only be done once if the first entry is a non-expired one.

It should happen more than *once* as there can be a lot of non-expired records (non-expired entries will be more than the expired ones at most times).

> If we use first_key_value we will have to remove the entry after that from expiries BTreeMap which adds a lot of overhead.

Instead of removing and inserting back unconditionally, iterating over references then removing when needed should be much quicker I think. Did you measure the performance difference between these two?

Reply:
> It should happen more than once as there can be a lot of non-expired records (non-expired entries will be more than the expired ones at most times).

But it's ordered so when the first case happens we end the while loop.

> Instead of removing and inserting back unconditionally, iterating over references then removing when needed should be much quicker I think. Did you measure the performance difference between these two?

I can measure it if needed :)

Reply:
> But it's ordered so when the first case happens we end the while loop.

I missed that part, yeah. I think we should add some explanation comments inside this function. It will look very confusing for other people.

Reply:
> I think we should add some explanation comments inside this function. It will look very confusing for other people.

Done.